File: java/queue-time/src/main/java/es/eriktorr/katas/QueueTimeCalculator.java
package es.eriktorr.katas;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class QueueTimeCalculator {

    public int queueTime(List<Integer> queue, int tillsCount) {
        Deque<Integer> deque = new ArrayDeque<>(queue);
        List<SelfCheckoutTill> tills = IntStream.range(0, tillsCount)
                .mapToObj(i -> aSelfCheckoutTill(deque))
                .collect(Collectors.toList());

        int queueTime = 0;

        do {
            int minQueueTime = tills.stream()
                    .mapToInt(SelfCheckoutTill::waitingTime)
                    .min().orElse(0);
            queueTime += minQueueTime;

            tills.forEach(till -> {
                till.advance(minQueueTime);
                if (till.waitingTime() == 0 && !deque.isEmpty()) till.waitingTime(deque.remove());
            });
        } while (!deque.isEmpty());

        queueTime += tills.stream()
                .mapToInt(SelfCheckoutTill::waitingTime)
                .max().orElse(0);

        return queueTime;
    }

    private SelfCheckoutTill aSelfCheckoutTill(Deque<Integer> deque) {
        SelfCheckoutTill till = new SelfCheckoutTill();
        if (!deque.isEmpty()) till.waitingTime(deque.remove());
        return till;
    }

    private static class SelfCheckoutTill {
        private int waitingTime = 0;

        void waitingTime(int waitingTime) {
            this.waitingTime = waitingTime;
        }

        int waitingTime() {
            return waitingTime;
        }

        void advance(int time) {
            waitingTime -= time;
        }
    }

}

File: kotlin/queue_time/src/main/kotlin/es/eriktorr/katas/QueueTimeCalculator.kt
package es.eriktorr.katas

import java.util.*

class QueueTimeCalculator {

    fun queueTime(queue: List<Int>, tillsCount: Int): Int {
        val deque = ArrayDeque<Int>(queue)
        val tills = generateSequence(0) { it + 1 }.take(tillsCount)
                .map { _ -> aSelfCheckoutTill(deque)}
                .toList()

        var queueTime = 0

        do {
            val minQueueTime = tills.minBy { it.waitingTime }!!.waitingTime
            queueTime += minQueueTime

            tills.forEach { it ->
                it.advance(minQueueTime)
                if (it.waitingTime == 0 && !deque.isEmpty()) it.waitingTime = deque.remove()
            }
        } while (!deque.isEmpty())

        queueTime += tills.maxBy { it.waitingTime }!!.waitingTime

        return queueTime
    }

    private fun aSelfCheckoutTill(deque: ArrayDeque<Int>): SelfCheckoutTill {
        return SelfCheckoutTill(if (!deque.isEmpty()) deque.remove() else 0)
    }

    private data class SelfCheckoutTill(var waitingTime: Int = 0) {
        fun advance(value: Int) {
            waitingTime -= value
        }
    }

}